jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ConditionalHandler.java

```diff
@@ -0,0 +1,735 @@
+        }
+
+        @Override
+        public String toString()
+        {
+            return String.format("%s@%x{%s}", getClass().getSimpleName(), hashCode(), _pathSpec);
+        }
+    }
+
+    /**
+     * An implementation of {@link ConditionalHandler} that will invoke {@link #nextHandle(Request, Response, Callback)} if
+     * the conditions are met, otherwise {@code false} is returned from {@link #doNotHandle(Request, Response, Callback)} and
+     * no further handling is done.
+     */
+    public static class Terminate extends ConditionalHandler
+    {
+        public Terminate()
+        {
+            super();
+        }
+
+        public Terminate(Handler handler)
+        {
+            super(handler);
+        }
+
+        @Override
+        protected boolean doHandle(Request request, Response response, Callback callback) throws Exception
+        {
+            return nextHandle(request, response, callback);
+        }
+
+        @Override
+        protected boolean doNotHandle(Request request, Response response, Callback callback) throws Exception
+        {
+            return false;
+        }
+    }
+
+    /**
+     * An implementation of {@link ConditionalHandler} that will invoke {@link #nextHandle(Request, Response, Callback)} if
+     * the conditions are met, otherwise {@link #doNotHandle(Request, Response, Callback)} sends a
+     * {@link HttpStatus#FORBIDDEN_403} response.
+     */
+    public static class Forbidden extends ConditionalHandler
+    {
+        public Forbidden()
+        {
+            super();
+        }
+
+        public Forbidden(Handler handler)
+        {
+            super(handler);
+        }
+
+        @Override
+        protected boolean doHandle(Request request, Response response, Callback callback) throws Exception
+        {
+            return nextHandle(request, response, callback);
+        }
+
+        @Override
+        protected boolean doNotHandle(Request request, Response response, Callback callback) throws Exception
+        {
+            Response.writeError(request, response, callback, HttpStatus.FORBIDDEN_403);
+            return true;
+        }
+    }
+
+    /**
+     * An implementation of {@link ConditionalHandler} that will invoke {@link #nextHandle(Request, Response, Callback)} if
+     * the conditions are met, otherwise {@link #doNotHandle(Request, Response, Callback)} will skip the next {@link Handler}
+     * by its {@link Singleton#getHandler() next Handler}.
+     */
+    public static class SkipNext extends ConditionalHandler
+    {
+        public SkipNext()
+        {
+            super();
+        }
+
+        public SkipNext(Handler handler)
+        {
+            super(handler);
+        }
+
+        @Override
+        protected boolean doHandle(Request request, Response response, Callback callback) throws Exception
+        {
+            Handler next = getHandler();
+            return next != null && next.handle(request, response, callback);
```

Comment:
Call `nextHandle()` here.

Reply:
yes, but now we get two stack frames between this method and the next handler.     I definitely prefer the manual inlining, but have removed under protest.